cmd/run.go

```diff
@@ -83,146 +125,193 @@ Default builder is pack.
-	Build string
+func newRunConfig(cmd *cobra.Command) (c runConfig) {
+	c = runConfig{
+		buildConfig: newBuildConfig(),
+		Build:       viper.GetString("build"),
+		Env:         viper.GetStringSlice("env"),
+		Container:   viper.GetBool("container"),
+	}
+	// NOTE: .Env should be viper.GetStringSlice, but this returns unparsed
+	// results and appears to be an open issue since 2017:
+	// https://github.com/spf13/viper/issues/380
+	var err error
+	if c.Env, err = cmd.Flags().GetStringArray("env"); err != nil {
+		fmt.Fprintf(cmd.OutOrStdout(), "error reading envs: %v", err)
+	}
+	return
+}
+
+// Configure the given function.  Updates a function struct with all
+// configurable values.  Note that the config alrady includes function's
+// current state, as they were passed through via flag defaults.
+func (c runConfig) Configure(f fn.Function) (fn.Function, error) {
+	var err error
+	f = c.buildConfig.Configure(f)
 
-	// Builder strategy if building
-	Builder string
+	f.Run.Envs, err = applyEnvs(f.Run.Envs, c.Env)
+	return f, err
 
-	// Registry for the build tag if building
-	Registry string
+	// The other members; build, path, and container; are not part of function
+	// state, so are not mentioned here in Configure.
 }
 
-func newRunConfig(cmd *cobra.Command) (cfg runConfig, err error) {
-	envToUpdate, envToRemove, err := envFromCmd(cmd)
-	if err != nil {
+func (c runConfig) Prompt() (runConfig, error) {
+	var err error
+
+	if c.buildConfig, err = c.buildConfig.Prompt(); err != nil {
+		return c, err
+	}
+
+	if !interactiveTerminal() || !c.Confirm {
+		return c, nil
+	}
+
+	// TODO:  prompt for additional settings here
+	return c, nil
+}
+
+func (c runConfig) Validate(cmd *cobra.Command) (err error) {
+	// Bubble
+	if err = c.buildConfig.Validate(); err != nil {
 		return
 	}
-	cfg = runConfig{
-		Build:       viper.GetString("build"),
-		Path:        viper.GetString("path"),
-		Verbose:     viper.GetBool("verbose"), // defined on root
-		Builder:     viper.GetString("builder"),
-		Registry:    viper.GetString("registry"),
-		EnvToUpdate: envToUpdate,
-		EnvToRemove: envToRemove,
+
+	// --build can be "auto"|true|false
+	if c.Build != "auto" {
+		if _, err := strconv.ParseBool(c.Build); err != nil {
+			return fmt.Errorf("unrecognized value for --build '%v'.  Accepts 'auto', 'true' or 'false' (or similarly truthy value)", c.Build)
+		}
+	}
+
+	// There is currently no local host runner implemented, so specifying
+	// --container=false should always return an informative error to the user
+	// such that they do not receive the rather cryptic "no runner defined"
+	// error from a Client instance which was instantiated with no runner.
+	// TODO: modify this check when the local host runner is available to
+	// only generate this error when --container==false && the --language is
+	// not yet implemented.
+	if !c.Container {
+		return errors.New("The ability to run functions outside of a container via 'func run' is coming soon.")
```

Comment:
Nit: I think convention is to start error messages with lower-case
```suggestion
		return errors.New("the ability to run functions outside of a container via 'func run' is coming soon.")
```